File: clients/typescript/src/subscriber.ts
/**
 * {@link WhdrSubscriber} — the batteries-included reconnect-and-resume loop.
 *
 * Implements the appendix §7 algorithm end-to-end: auth → welcome → subscribe
 * with `replay.after_seq = cursor` → dedup by `id` and `seq <= cursor` →
 * advance/persist the cursor only after a successful handle → recover from
 * `lagged` / dropped sockets by reconnecting from the cursor → surface
 * `replay_gap` → treat `revoked` as fatal and `shutdown` as a backoff
 * reconnect → exponential backoff with jitter → ignore unknown frames.
 */
import type { ClientOptions } from "ws";

import { Backoff, DEFAULT_BACKOFF, type BackoffPolicy } from "./backoff.js";
import { Connection } from "./connection.js";
import {
  MemoryCursorStore,
  type CursorStore,
} from "./cursor.js";
import {
  CursorStoreError,
  HandlerError,
  RevokedError,
  WhdrError,
} from "./errors.js";
import type { EventFrame } from "./frames.js";
import { ResumeState } from "./resume.js";

/** Decode a standard-base64 payload to raw bytes. */
export function decodePayload(payloadB64: string): Uint8Array {
  return new Uint8Array(Buffer.from(payloadB64, "base64"));
}

/**
 * A delivered event handed to a {@link Handler}. The kit has already
 * de-duplicated by `id`/`seq`, so a handler sees each event at most once. Dedup
 * key is `id`; order by `seq`, not `ts_ms`.
 */
export interface DeliveredEvent {
  readonly id: string;
  readonly seq: number;
  readonly ts_ms: number;
  readonly channel: string;
  readonly payload_b64: string;
  /** Decode `payload_b64` to raw bytes. */
  payload(): Uint8Array;
}

function makeDeliveredEvent(frame: EventFrame): DeliveredEvent {
  return {
    id: frame.id,
    seq: frame.seq,
    ts_ms: frame.ts_ms,
    channel: frame.channel,
    payload_b64: frame.payload_b64,
    payload: () => decodePayload(frame.payload_b64),
  };
}

/**
 * Callbacks for the {@link WhdrSubscriber.run} loop. Only {@link Handler.onEvent}
 * is required. **Throwing from any hook is fatal**: `run` stops and rethrows a
 * {@link HandlerError}. The cursor advances (and persists) only *after*
 * `onEvent` resolves — at-least-once delivery.
 */
export interface Handler {
  /** Handle a delivered event. On success the cursor advances to `event.seq`. */
  onEvent(event: DeliveredEvent): void | Promise<void>;
  /** A replay window finished at `throughSeq`; live frames follow. */
  onReplayed?(throughSeq: number): void | Promise<void>;
  /**
   * Explicit data-loss signal: events in `(fromSeq, earliestSeq)` were pruned
   * before this subscriber resumed. Reconcile out-of-band if you must.
   */
  onReplayGap?(fromSeq: number, earliestSeq: number): void | Promise<void>;
  /** The server evicted `dropped` events; the kit reconnects and replays. */
  onLagged?(dropped: number): void | Promise<void>;
  /** A `replay` request was refused (durability disabled); live still works. */
  onReplayUnavailable?(msg: string): void | Promise<void>;
}

/** Construction options for a {@link WhdrSubscriber}. */
export interface WhdrSubscriberOptions {
  /** `/subscribe` endpoint, e.g. `ws://127.0.0.1:8788/subscribe`. */
  url: string;
  /** `tok_…` subscriber token minted by the operator. */
  token: string;
  /** NATS-style channel patterns, e.g. `["github.>"]`. */
  patterns: string[];
  /** Initial resume cursor (default 0 = replay from retention start). Ignored when `cursorStore` is set. */
  cursor?: number;
  /** Cursor-persistence hook for at-least-once across restarts. */
  cursorStore?: CursorStore;
  /** Backoff policy overrides (merged over the defaults). */
  backoff?: Partial<BackoffPolicy>;
  /** Recent-`id` dedup window size (default 8192). */
  dedupCapacity?: number;
  /** Extra `ws` client options (e.g. TLS). */
  wsOptions?: ClientOptions;
  /** Jitter source, injectable for tests (default `Math.random`). */
  rand?: () => number;
  /** Sleep implementation, injectable for tests (default abortable `setTimeout`). */
  sleep?: (ms: number, signal?: AbortSignal) => Promise<void>;
}

/** Options for a single {@link WhdrSubscriber.run} invocation. */
export interface RunOptions {
  /** Abort to stop the loop cleanly (closes the socket, no reconnect). */
  signal?: AbortSignal;
}

/** Options for {@link WhdrSubscriber.stream}: an abort signal plus signal hooks. */
export interface StreamOptions extends RunOptions {
  onReplayed?(throughSeq: number): void | Promise<void>;
  onReplayGap?(fromSeq: number, earliestSeq: number): void | Promise<void>;
  onLagged?(dropped: number): void | Promise<void>;
  onReplayUnavailable?(msg: string): void | Promise<void>;
}

function defaultSleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) return resolve();
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = (): void => {
      clearTimeout(timer);
      resolve();
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

export class WhdrSubscriber {
  readonly #url: string;
  readonly #token: string;
  readonly #patterns: string[];
  readonly #cursorStore: CursorStore;
  readonly #backoffPolicy: BackoffPolicy;
  readonly #dedupCapacity: number;
  readonly #wsOptions: ClientOptions;
  readonly #rand: () => number;
  readonly #sleep: (ms: number, signal?: AbortSignal) => Promise<void>;

  constructor(options: WhdrSubscriberOptions) {
    this.#url = options.url;
    this.#token = options.token;
    this.#patterns = [...options.patterns];
    this.#cursorStore =
      options.cursorStore ?? new MemoryCursorStore(options.cursor ?? 0);
    this.#backoffPolicy = { ...DEFAULT_BACKOFF, ...options.backoff };
    this.#dedupCapacity = Math.max(1, options.dedupCapacity ?? 8192);
    this.#wsOptions = options.wsOptions ?? {};
    this.#rand = options.rand ?? Math.random;
    this.#sleep = options.sleep ?? defaultSleep;
  }

  /**
   * Connect, authenticate, and subscribe once with the configured patterns and
   * cursor. Returns a ready {@link Connection} for bespoke loops; most callers
   * want {@link WhdrSubscriber.run}.
   */
  async connect(): Promise<Connection> {
    const cursor = await this.#cursorStore.load();
    const conn = await Connection.connect(this.#url, this.#token, this.#wsOptions);
    await conn.subscribe(this.#patterns, cursor);
    return conn;
  }

  /**
   * Run the full reconnect-and-resume loop, driving `handler`. Loops forever,
   * reconnecting with backoff after a transient failure. Returns when `signal`
   * aborts; rejects only on a **fatal** error (auth/revoked/handler/cursor).
   */
  async run(handler: Handler, options: RunOptions = {}): Promise<void> {
    const signal = options.signal;
    const initial = await this.#cursorStore.load();
    const resume = new ResumeState(initial, this.#dedupCapacity);
    const backoff = new Backoff(this.#backoffPolicy, this.#rand);

    while (!signal?.aborted) {
      try {
        await this.#runSession(handler, resume, backoff, signal);
      } catch (err) {
        if (err instanceof WhdrError && err.fatal) throw err;
        if (signal?.aborted) return;
        // Transient: fall through to backoff + reconnect.
      }
      if (signal?.aborted) return;
      await this.#sleep(backoff.nextDelayMs(), signal);
    }
  }

  /**
   * Async-iterator event API. Yields de-duplicated {@link DeliveredEvent}s while
   * the reconnect-and-resume loop runs in the background. Non-event signals are
   * delivered via the optional hooks in `options`. Call `return()` on the
   * iterator (or abort `options.signal`) to stop.
   */
  stream(options: StreamOptions = {}): AsyncIterableIterator<DeliveredEvent> {
    const controller = new AbortController();
    if (options.signal) {
      if (options.signal.aborted) controller.abort();
      else options.signal.addEventListener("abort", () => controller.abort(), { once: true });
    }

    const queue = new EventQueue();
    const handler: Handler = {
      onEvent: (event) => queue.push(event),
      ...(options.onReplayed ? { onReplayed: options.onReplayed } : {}),
      ...(options.onReplayGap ? { onReplayGap: options.onReplayGap } : {}),
      ...(options.onLagged ? { onLagged: options.onLagged } : {}),
      ...(options.onReplayUnavailable
        ? { onReplayUnavailable: options.onReplayUnavailable }
        : {}),
    };

    this.run(handler, { signal: controller.signal }).then(
      () => queue.finish(),
      (err) => queue.fail(err),
    );

    const iterator: AsyncIterableIterator<DeliveredEvent> = {
      next: () => queue.next(),
      return: async () => {
        controller.abort();
        queue.finish();
        return { done: true, value: undefined };
      },
      [Symbol.asyncIterator]() {
        return iterator;
      },
    };
    return iterator;
  }

  async #runSession(
    handler: Handler,
    resume: ResumeState,
    backoff: Backoff,
    signal: AbortSignal | undefined,
  ): Promise<void> {
    const conn = await Connection.connect(this.#url, this.#token, this.#wsOptions);
    const onAbort = (): void => conn.close();
    signal?.addEventListener("abort", onAbort, { once: true });
    try {
      // Connected: reset backoff so a later drop reconnects fast.
      backoff.reset();
      await conn.subscribe(this.#patterns, resume.cursor());

      for (;;) {
        const frame = await conn.recv();
        switch (frame.type) {
          case "event": {
            if (resume.shouldProcess(frame.id, frame.seq)) {
              const event = makeDeliveredEvent(frame);
              await callHook(() => handler.onEvent(event));
              resume.record(frame.id, frame.seq);
              try {
                await this.#cursorStore.save(resume.cursor());
              } catch (err) {
                throw new CursorStoreError(err);
              }
            }
            break;
          }
          case "replayed":
            await callHook(() => handler.onReplayed?.(frame.through_seq));
            break;
          case "replay_gap":
            await callHook(() =>
              handler.onReplayGap?.(frame.from_seq, frame.earliest_seq),
            );
            break;
          case "lagged":
            await callHook(() => handler.onLagged?.(frame.dropped));
            // Recover by reconnecting and replaying from the cursor.
            return;
          case "error":
            if (frame.op === "replay") {
              await callHook(() => handler.onReplayUnavailable?.(frame.msg));
            }
            // Other errors (e.g. bad pattern) are non-fatal; keep the connection.
            break;
          case "closing":
            if (frame.reason === "revoked") throw new RevokedError();
            // shutdown: reconnect with backoff.
            return;
          case "welcome":
          case "ok":
          case "pong":
            // Nothing to do.
            break;
        }
      }
    } finally {
      signal?.removeEventListener("abort", onAbort);
      conn.close();
    }
  }
}

/** Await an optional hook, wrapping any throw as a fatal {@link HandlerError}. */
async function callHook(fn: () => void | Promise<void>): Promise<void> {
  try {
    await fn();
  } catch (err) {
    if (err instanceof WhdrError) throw err;
    throw new HandlerError(err);
  }
}

/** A single-consumer async queue bridging the callback loop to an iterator. */
class EventQueue {
  readonly #items: DeliveredEvent[] = [];
  #resolve: ((result: IteratorResult<DeliveredEvent>) => void) | null = null;
  #reject: ((err: unknown) => void) | null = null;
  #done = false;
  #error: unknown = null;

  push(event: DeliveredEvent): void {
    if (this.#resolve) {
      const resolve = this.#resolve;
      this.#resolve = null;
      this.#reject = null;
      resolve({ done: false, value: event });
    } else {
      this.#items.push(event);
    }
  }

  finish(): void {
    this.#done = true;
    if (this.#resolve) {
      const resolve = this.#resolve;
      this.#resolve = null;
      this.#reject = null;
      resolve({ done: true, value: undefined });
    }
  }

  fail(err: unknown): void {
    this.#error = err;
    if (this.#reject) {
      const reject = this.#reject;
      this.#resolve = null;
      this.#reject = null;
      reject(err);
    }
  }

  next(): Promise<IteratorResult<DeliveredEvent>> {
    const item = this.#items.shift();
    if (item !== undefined) return Promise.resolve({ done: false, value: item });
    if (this.#error !== null) return Promise.reject(this.#error);
    if (this.#done) return Promise.resolve({ done: true, value: undefined });
    return new Promise((resolve, reject) => {
      this.#resolve = resolve;
      this.#reject = reject;
    });
  }
}

File: clients/python/src/whdr_sub_kit/_subscriber.py
"""The :class:`Subscriber` — the batteries-included reconnect-and-resume client.

Implements the appendix §7 algorithm: auth -> welcome -> subscribe with
``replay.after_seq = cursor`` -> dedup by ``id`` / ``seq`` -> advance the cursor
after each successful handle -> recover from ``lagged`` / disconnects by
reconnecting from the cursor -> surface ``replay_gap`` -> treat ``revoked`` as
fatal and ``shutdown`` as a backoff reconnect.
"""

from __future__ import annotations

import asyncio
import inspect
import logging
from collections.abc import AsyncIterator, Awaitable, Callable, Iterable
from typing import Any, Union

from ._backoff import Backoff, BackoffPolicy
from ._connection import Connection
from ._cursor import CursorStore, MemoryCursorStore, ResumeState
from ._errors import (
    CursorStoreError,
    HandlerError,
    RevokedError,
    SubscriberError,
)
from ._frames import (
    Closing,
    DeliveredEvent,
    ErrorFrame,
    Lagged,
    Replayed,
    ReplayGap,
    ServerFrame,
)

logger = logging.getLogger("whdr_sub_kit")

__all__ = ["Subscriber", "Handler", "EventCallback", "SignalCallback"]

#: An async or sync callable handling a single delivered event.
EventCallback = Callable[[DeliveredEvent], Union[Awaitable[None], None]]
#: An async or sync callable for a signal frame (replayed / lagged / ...).
SignalCallback = Callable[..., Union[Awaitable[None], None]]


class Handler:
    """Base class for the :meth:`Subscriber.run` handler.

    Override :meth:`on_event` (required); the signal hooks default to logging
    (for ``replay_gap``) or nothing. Raising from any hook is **fatal**: the run
    loop stops and raises :class:`~whdr_sub_kit.HandlerError`. The cursor is
    advanced (and persisted) only *after* :meth:`on_event` returns, giving
    at-least-once delivery.
    """

    async def on_event(self, event: DeliveredEvent) -> None:
        """Handle a delivered event (de-duplicated: called at most once)."""
        raise NotImplementedError

    async def on_replayed(self, through_seq: int) -> None:
        """A replay window finished; live frames follow. Default: no-op."""

    async def on_replay_gap(self, from_seq: int, earliest_seq: int) -> None:
        """Explicit data-loss signal: events in ``(from_seq, earliest_seq)`` were
        pruned before this subscriber resumed. Default: logs a warning."""
        logger.warning(
            "replay_gap: events (%d, %d) were pruned before this subscriber resumed",
            from_seq,
            earliest_seq,
        )

    async def on_lagged(self, dropped: int) -> None:
        """The server evicted ``dropped`` events; the kit reconnects and replays
        from the cursor to recover. Default: no-op."""

    async def on_replay_unavailable(self, msg: str) -> None:
        """A ``replay`` request was refused because durable delivery is disabled;
        live delivery still works. Default: no-op."""


async def _maybe_await(value: Awaitable[None] | None) -> None:
    if inspect.isawaitable(value):
        await value


class _Signals:
    """Signal-frame sinks built from the :class:`Subscriber` constructor
    callbacks, with the reference defaults (``replay_gap`` logs; others no-op).

    Shares its method names with :class:`_HandlerHooks` so both satisfy the
    control-frame dispatcher's interface.
    """

    __slots__ = ("_replayed", "_replay_gap", "_lagged", "_replay_unavailable")

    def __init__(
        self,
        on_replayed: SignalCallback | None,
        on_replay_gap: SignalCallback | None,
        on_lagged: SignalCallback | None,
        on_replay_unavailable: SignalCallback | None,
    ) -> None:
        self._replayed = on_replayed
        self._replay_gap = on_replay_gap
        self._lagged = on_lagged
        self._replay_unavailable = on_replay_unavailable

    async def replayed(self, through_seq: int) -> None:
        if self._replayed is not None:
            await _maybe_await(self._replayed(through_seq))

    async def replay_gap(self, from_seq: int, earliest_seq: int) -> None:
        if self._replay_gap is not None:
            await _maybe_await(self._replay_gap(from_seq, earliest_seq))
        else:
            logger.warning(
                "replay_gap: events (%d, %d) were pruned before this subscriber resumed",
                from_seq,
                earliest_seq,
            )

    async def lagged(self, dropped: int) -> None:
        if self._lagged is not None:
            await _maybe_await(self._lagged(dropped))

    async def replay_unavailable(self, msg: str) -> None:
        if self._replay_unavailable is not None:
            await _maybe_await(self._replay_unavailable(msg))


class _HandlerHooks:
    """Adapts a :class:`Handler` to the control-frame dispatcher interface."""

    __slots__ = ("_handler",)

    def __init__(self, handler: Handler) -> None:
        self._handler = handler

    async def replayed(self, through_seq: int) -> None:
        await self._handler.on_replayed(through_seq)

    async def replay_gap(self, from_seq: int, earliest_seq: int) -> None:
        await self._handler.on_replay_gap(from_seq, earliest_seq)

    async def lagged(self, dropped: int) -> None:
        await self._handler.on_lagged(dropped)

    async def replay_unavailable(self, msg: str) -> None:
        await self._handler.on_replay_unavailable(msg)


class _FnHandler(Handler):
    """Wraps a bare event callable as a :class:`Handler`, routing signal frames
    to the subscriber's constructor callbacks."""

    def __init__(self, fn: EventCallback, signals: _Signals) -> None:
        self._fn = fn
        self._signals = signals

    async def on_event(self, event: DeliveredEvent) -> None:
        await _maybe_await(self._fn(event))

    async def on_replayed(self, through_seq: int) -> None:
        await self._signals.replayed(through_seq)

    async def on_replay_gap(self, from_seq: int, earliest_seq: int) -> None:
        await self._signals.replay_gap(from_seq, earliest_seq)

    async def on_lagged(self, dropped: int) -> None:
        await self._signals.lagged(dropped)

    async def on_replay_unavailable(self, msg: str) -> None:
        await self._signals.replay_unavailable(msg)


# Interface shared by _Signals and _HandlerHooks.
_ControlHooks = Union[_Signals, _HandlerHooks]


async def _dispatch_control(frame: ServerFrame, hooks: _ControlHooks) -> str | None:
    """Handle a non-event frame.

    Returns ``"reconnect"`` when the session should end and resume from the
    cursor (``lagged``, or a ``closing`` that is not ``revoked``); ``None`` to
    keep reading. Raises :class:`RevokedError` for a ``revoked`` close.
    """
    if isinstance(frame, Replayed):
        await hooks.replayed(frame.through_seq)
        return None
    if isinstance(frame, ReplayGap):
        await hooks.replay_gap(frame.from_seq, frame.earliest_seq)
        return None
    if isinstance(frame, Lagged):
        await hooks.lagged(frame.dropped)
        return "reconnect"
    if isinstance(frame, ErrorFrame):
        if frame.op == "replay":
            logger.warning("replay refused (durability disabled): %s", frame.msg)
            await hooks.replay_unavailable(frame.msg)
        else:
            logger.warning("server error frame op=%s msg=%s", frame.op, frame.msg)
        return None
    if isinstance(frame, Closing):
        if frame.reason == "revoked":
            raise RevokedError()
        # "shutdown" (or any other reason): reconnect with backoff.
        return "reconnect"
    # Welcome (unexpected repeat), Ok, Pong: nothing to do.
    return None


class Subscriber:
    """A configured, self-driving whdr subscriber.

    Two ways to consume events, both implementing the full reconnect-and-resume
    algorithm (dedup, cursor advance, backoff, ``lagged``/``shutdown`` recovery,
    ``revoked`` = fatal):

    - ``async for event in sub.events(): ...`` — the typed event stream. The
      cursor advances (and is persisted) once your loop body finishes an event.
    - ``await sub.run(handler)`` — hand a :class:`Handler` (or a bare async
      callable) and the kit calls it per event, advancing the cursor after each
      successful call.

    Signal frames (``replayed``, ``replay_gap``, ``lagged``,
    ``replay_unavailable``) are surfaced via the optional constructor callbacks
    for :meth:`events`, or via the :class:`Handler`'s hooks for :meth:`run`.
    """

    def __init__(
        self,
        url: str,
        token: str,
        *,
        patterns: Iterable[str] | str,
        cursor: int = 0,
        cursor_store: CursorStore | None = None,
        backoff: BackoffPolicy | None = None,
        dedup_capacity: int = 8192,
        on_replayed: SignalCallback | None = None,
        on_replay_gap: SignalCallback | None = None,
        on_lagged: SignalCallback | None = None,
        on_replay_unavailable: SignalCallback | None = None,
        open_timeout: float | None = 10.0,
        connect_kwargs: dict[str, Any] | None = None,
    ) -> None:
        self._url = url
        self._token = token
        self._patterns = [patterns] if isinstance(patterns, str) else list(patterns)
        self._store: CursorStore = cursor_store or MemoryCursorStore(cursor)
        self._backoff = backoff or BackoffPolicy()
        self._dedup_capacity = max(dedup_capacity, 1)
        self._signals = _Signals(
            on_replayed, on_replay_gap, on_lagged, on_replay_unavailable
        )
        opts: dict[str, Any] = dict(connect_kwargs or {})
        if open_timeout is not None:
            opts.setdefault("open_timeout", open_timeout)
        self._opts = opts

    # -- construction helpers -------------------------------------------------

    def _connection(self) -> Connection:
        return Connection(self._url, self._token, **self._opts)

    def _normalize_handler(self, handler: Handler | EventCallback) -> Handler:
        if isinstance(handler, Handler):
            return handler
        if callable(handler):
            return _FnHandler(handler, self._signals)
        raise TypeError("handler must be a Handler or an event callable")

    async def _load_cursor(self) -> int:
        try:
            return await self._store.load()
        except Exception as err:  # noqa: BLE001 — cursor store is a user hook
            raise CursorStoreError(err) from err

    async def _save_cursor(self, cursor: int) -> None:
        try:
            await self._store.save(cursor)
        except Exception as err:  # noqa: BLE001 — cursor store is a user hook
            raise CursorStoreError(err) from err

    # -- run(handler) ---------------------------------------------------------

    async def run(self, handler: Handler | EventCallback) -> None:
        """Run the full reconnect-and-resume loop, driving ``handler``.

        Loops forever, reconnecting with exponential backoff after a transient
        failure (dropped socket, server ``shutdown``, ``lagged`` eviction).
        Returns only if the inner loop is broken by cancellation; raises on a
        **fatal** error (revoked/absent token, handler failure, cursor-store
        failure).
        """
        h = self._normalize_handler(handler)
        hooks = _HandlerHooks(h)
        resume = ResumeState(await self._load_cursor(), self._dedup_capacity)
        backoff = self._backoff.start()
        while True:
            try:
                await self._run_session(h, hooks, resume, backoff)
                logger.info("subscriber session ended; reconnecting")
            except SubscriberError as err:
                if err.is_fatal():
                    raise
                logger.warning("subscriber session error; reconnecting: %s", err)
            await asyncio.sleep(backoff.next_delay())

    async def _run_session(
        self,
        handler: Handler,
        hooks: _HandlerHooks,
        resume: ResumeState,
        backoff: Backoff,
    ) -> None:
        async with self._connection() as conn:
            # Connected: reset backoff so a later drop reconnects fast.
            backoff.reset()
            await conn.subscribe(self._patterns, resume.cursor)
            while True:
                frame = await conn.recv()
                if isinstance(frame, DeliveredEvent):
                    if resume.should_process(frame.id, frame.seq):
                        try:
                            await handler.on_event(frame)
                        except asyncio.CancelledError:
                            raise
                        except Exception as err:  # noqa: BLE001
                            raise HandlerError(err) from err
                        resume.record(frame.id, frame.seq)
                        await self._save_cursor(resume.cursor)
                else:
                    if await _dispatch_control(frame, hooks) == "reconnect":
                        return

    # -- events() iterator ----------------------------------------------------

    async def events(self) -> AsyncIterator[DeliveredEvent]:
        """Yield de-duplicated events, reconnecting and resuming forever.

        The cursor advances (and is persisted) after your loop body finishes
        processing each yielded event, giving at-least-once delivery. Signal
        frames go to the constructor callbacks (``replay_gap`` logs by default).
        A fatal error (revoked token, cursor-store failure) propagates out of
        the iterator; transient failures reconnect with backoff.
        """
        resume = ResumeState(await self._load_cursor(), self._dedup_capacity)
        backoff = self._backoff.start()
        hooks = self._signals
        while True:
            try:
                async with self._connection() as conn:
                    backoff.reset()
                    await conn.subscribe(self._patterns, resume.cursor)
                    reconnect = False
                    while not reconnect:
                        frame = await conn.recv()
                        if isinstance(frame, DeliveredEvent):
                            if resume.should_process(frame.id, frame.seq):
                                yield frame
                                resume.record(frame.id, frame.seq)
                                await self._save_cursor(resume.cursor)
                        elif await _dispatch_control(frame, hooks) == "reconnect":
                            reconnect = True
            except SubscriberError as err:
                if err.is_fatal():
                    raise
                logger.warning("subscriber session error; reconnecting: %s", err)
            await asyncio.sleep(backoff.next_delay())
